Feature: document comments

tim-vogel opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

<!--
Please include a summary of the change and which issue is fixed (if any) and any relevant motivation / context. List any dependencies that are required for this change. If appropriate, please include an explanation of how your proposed change can be tested. Screenshots and / or videos can also be helpful if appropriate.
-->

**new Feature list:**
- [x] - Tab for displaying comments in the document detail view.
- [x] - List all comments for a document in descending order (newest comments first)
- [x] - add new comments
- [x] - delete comments
- [ ] - ~~full text search over comments~~
- [x] -  integration of comments into the standard import and export function of paperless
- [x] - activate / deactivate the comment function (only the view is hidden or shown)
- [x] - new environment variable ```PAPERLESS_COMMENTS_ENABLED``` -> [true: enabled | false: disabled]
- [x] - new service to retrieve backend environment variables. 

By default the comment function is disabled. To enable it, the environment variable ```PAPERLESS_COMMENTS_ENABLED``` must be set to ```true```

A possible full text search for the comments would have to be added by an experienced programmer.

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [ ] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

tooomm committed:
typo

tim-vogel committed:
add comment function

shamoon commented:
Thanks, we’ll take a look. One immediate thing: settings are already saved to the database, I think this should use that rather than a new Env var service.

shamoon commented:
Changing this to draft, at least frontend code needs some work, so does the UI. Also (as you noted) the `pre-commit` hooks weren't run on the code so the style checks are failing. 

Definitely appreciate the work!

shamoon committed:
Use frontend settings instead of env variable

shamoon committed:
Remove faux testing file

shamoon committed:
Fix comment service method signature

and run prettier over file

shamoon committed:
Remove faux test file

shamoon committed:
Missed PAPERLESS_COMMENTS_ENABLED code

shamoon committed:
fix migration dependency

and run code style

shamoon committed:
Refactor comment UI code

And run prettier

shamoon committed:
python code style

shamoon committed:
Rework comment UI

Add error popups, text field validation, move form, move comment header to footer, updated styling

shamoon committed:
Cleanup frontend comment service & other code

[ci skip]

shamoon commented:
Ok! Im happier with the frontend code 😅, please take a look @tim-vogel and let me know if you have any concerns. Screenshots of new styling are below for the curious

TODO (AFAIK):

- [ ] backend checks are failing
- [ ] frontend tests
- [ ] search?

<img width="48%" alt="Screen Shot 2022-08-07 at 9 37 13 PM" src="https://user-images.githubusercontent.com/4887959/183344943-d8172411-0b5a-4c51-a9df-d155a9f73b96.png"> <img width="48%" alt="Screen Shot 2022-08-07 at 9 36 59 PM" src="https://user-images.githubusercontent.com/4887959/183344950-84bb68a4-b979-4b79-a9ce-4421edbbae06.png">

shamoon committed:
Update messages.xlf

tim-vogel commented:
@shamoon 

The new design looks great. What I miss on the screenshots is the possibility to delete a comment. 

Moving the setting to activate the comment function to the GUI is absolutely ok for me. The main thing is that the users who don't want this feature can disable it. 

With the open point of the search from the todo, I'm curious about the source code if someone implements this. I did not know how to implement this properly.

I will try to test the revised source code tonight.

Thanks for your revisions and corrections to make a clean source code out of my design. I'm sure it took you some time to correct everything.

shamoon commented:
> The new design looks great. What I miss on the screenshots is the possibility to delete a comment.

The delete button shows only when you hover over a comment! =)

tim-vogel commented:
I really like the feature with the fade-in delete button. The design is also very good. But what bothers me is that new comments are entered below the comment list.  For longer comment lists, you always have to scroll to the end of the comment list to enter a new comment. I think it would be better if the input field for new comments was placed at the top.

shamoon commented:
> I think it would be better if the input field for new comments was placed at the top.

I debated this a bit, I think there are pros / cons to each choice. Like if you think users will be coming here more often to view comments or to add a new one, that sorta determines where it goes =) Visually I prefer at the bottom but open to discussion on this of course. I can imagine some visual tweaks that would help if others feel strongly it be moved

mtettke123 commented:
+1 for top and sorting comments newest (top) to oldest (bottom)

tim-vogel commented:
Maybe it can be made adjustable. By means of CSS the order should be configurable. 

https://www.w3schools.com/cssref/css3_pr_order.asp

So the current view can remain as default. If the user wants to reverse the position, the order can be swapped with a CSS class.

Then another bool option in the settings and a CSS class would have to be created.

shamoon commented:
Two separate things:

- Sorting of comments. Currently newest on top, dont think we need a toggle for this imho unless many feel so.
- "Add a comment" box on top vs bottom. Making this toggle-able just feels weird to me, like indecision. Let's get some input and try to reach a consensus. Ultimately there will probably always be opinions on both sides! =)

tim-vogel commented:
Getting opinions is always a good idea ;-). Offering an option to leave the decision to the user I do not feel as indecisive. Rather, it offers flexibility and each user can set their own preferences. As you have already written, there will be lovers for each variant. 

But let the users decide how they want it ;-)

tim-vogel commented:
I consider a setting for sorting the comments to be superfluous and not purposeful.

enerschnutz commented:
> +1 for top and sorting comments newest (top) to oldest (bottom)

I second this

torwag commented:
Talking about UX, would it be a good idea to have a symbol for comments in the matrix view and adding the newest comment in the list-view? So users do not have to go into the detail view to read the latest comment.

Solverz-0 commented:
Will comments be included in the Document exporter when exporting/backing up paperless-ngx?

stumpylog commented:
> Will comments be included in the Document exporter when exporting/backing up paperless-ngx?

They will.

tim-vogel commented:
> Talking about UX, would it be a good idea to have a symbol for comments in the matrix view and adding the newest comment in the list-view? So users do not have to go into the detail view to read the latest comment.

I think this is a very good idea. However, there will be some different views on this again. For this reason, I think that we should first discuss this extension in a discussion. In addition, we should get the comment function into a release first. Extensions can then be added later.

NetHans commented:
The topic has been somewhat forgotten.
How do we want to continue at this point?

shamoon commented:
Backend tests are still failing, needs backend review. (Also one open question about placement of add comment box)

stumpylog commented:
The backend test failures need to be updated for a new item in the manifest.  And the export of comment and their contentneeds to be checked in tests.  Looking at the other code atm.

torwag commented:
Back to the UI decision,  I feel that both add new box and sorting order are mutual connected.  You want the "add new"  box close to the newest comment,  so that adding a new comment it would be listed right there for you to see.  It would be odd to add a new comment only to scroll down (or up)  a long list of comments to see the one you just added.

shamoon commented:
> Back to the UI decision, I feel that both, "add new" box and sorting order are mutual connected.

Yes this was exactly my reasoning. Clicking “add comment” and then it shows at the opposite end of the list seemed wrong. Since newest comment on top made more sense that’s where I put the form.

shamoon committed:
Move add comment to top of comments list

shamoon commented:
Ok, Im good with moving the form to the top, made the change.

stumpylog commented:
I've fixed the backend testing, added a basic verification that comments make it into the JSON, as well as fixing a minor issue with the format of the export.

shamoon commented:
Uh oh I just realized this is targeted to the `main` branch but should be to `dev`. Sorry for not noting earlier. I will change it but likely needs a rebase

stumpylog commented:
whoops, that's lost the backend testing and the export fix.

shamoon commented:
> whoops, that's lost the backend testing and the export fix.

my bad, hope you have it locally. Im adding frontend tests too

shamoon committed:
Add frontend tests for comments

shamoon committed:
Basic verification of Comment export & exporter comment tuple fix

From oprhaned commits https://github.com/paperless-ngx/paperless-ngx/commit/b1855a4b7af689d0a7c7f18bf7ef513967da269f
https://github.com/paperless-ngx/paperless-ngx/commit/53f21574fd8af0f3561c12f709a14980f8f1cb7f

Co-Authored-By: Trenton Holmes <797416+stumpylog@users.noreply.github.com>

shamoon commented:
Ok added frontend tests and added back your lost commits @stumpylog , not sure why tests are still failing?

stumpylog committed:
Merges the migration branches

stumpylog committed:
Starts on implementing tests for the new API

stumpylog committed:
Finalizes testing of the delete comment

stumpylog commented:
Ok, that should fix all the backend and such.

One thing I notice, when deleting a comment, I'm taken back to the overall documents screen, instead of the detail/comments tab where I deleted the comment from.  Is that the expected behavior?  My idea would be it simply removes the comment, maybe refreshing the page or data to show it gone?

shamoon commented:
> One thing I notice, when deleting a comment, I'm taken back to the overall documents screen, instead of the detail/comments tab where I deleted the comment from.

Oh weird, thats a regression of some kind, let me look into it, thanks for all the backend test stuff!

shamoon committed:
Fix CodeQL warnings

shamoon reviewed (approved):
Phew! ok this should be good-to-go...

Maybe search can be added down the road =)

stumpylog commented on src/documents/views.py:
```diff
@@ -379,6 +380,63 @@
+            try:
+                return Response(self.getComments(doc))
+            except Exception as e:
+                return Response({"error": str(e)})
```
I'm inclined to dismiss these, though they'd be easy to silence by only logging the exception and returning the generic error dictionary.  Thoughts?

shamoon replied:
Yea totally defer to you. I don’t know how useful the actual errors would be in practice. I changed this to log the full error and show a more generic output to UI, mostly to satisfy this QL warning. Again up to you of course if you want to change more!

stumpylog replied:
Nope, hadn't noticed your changes when I made this, that's exactly what I had in mind.

shamoon commented:
Thanks all, we can continue testing this feature in `dev` prior to 1.8.1

christofkac commented:
I've tested this feature now and have seen, that the comments are not exported in the same directory as the documents but rather "only" in the manifest.json.
As the search option is not yet available I thought it would be good to export the comments also in e.g. a text or JSON file next to the document. With this, you could use a normal file manager to search for those.
What do you think?
